### PR TITLE
Per-backend locale

### PR DESCRIPTION
**Pull request by whatyouhide**
Implementation of what was discussed in #48. Wdyt? /cc @lexmag @josevalim

### CODE REVIEW

File: lib/gettext.ex [hunk 1]
```diff
@@ -386,50 +385,52 @@ defmodule Gettext do
...
-    if locale = Process.get(__MODULE__) do
+  @spec get_locale(backend) :: locale
+  def get_locale(backend) do
+    if locale = Process.get(backend) do
       locale
     else
-      default_locale = Application.get_env(:gettext, :default_locale)
-      Process.put(__MODULE__, default_locale)
+      backend_config = Application.get_env(backend.__gettext__(:otp_app), backend, [])
+      default_locale = backend_config[:default_locale] || "en"
```
Comment (lexmag) [inline lib/gettext.ex]: I think `Keyword.get/3` would be better here, as we don't silently rewrite `default_locale: nil` case.

Comment (whatyouhide) [inline lib/gettext.ex]: Makes sense, pushed the change you suggested.

File: lib/gettext.ex [hunk 2]
```diff
@@ -246,7 +247,7 @@ defmodule Gettext do
 
   the `ngettext` macro can be used like this:
 
-      Gettext.locale "it"
+      Gettext.locale(MyApp.Gettext, "it")
```
Comment (lexmag) [inline lib/gettext.ex]: `put_locale`

File: lib/gettext.ex [hunk 3]
```diff
@@ -281,7 +282,7 @@ defmodule Gettext do
 
   For example:
 
-      Gettext.locale "foo"
+      Gettext.locale(MyApp.Gettext, "foo")
```
Comment (lexmag) [inline lib/gettext.ex]: `put_locale`

**Comment by lexmag** [comment]
I really feel that per backend `with_locale` will be handy.
As today we expect users to `import MyApp.Gettext`, but not having `with_locale` imported may defeat the the value of such import I think.
For instance:
```elixir
  import MyApp.Gettext

  defp translate_header(locale, bindings) do
    Gettext.with_locale(MyApp.Gettext, locale, fn() ->
      gettext <some text>, bindings
    end)
  end

  defp translate_body(locale) do
    Gettext.with_locale(MyApp.Gettext, locale, fn() ->
      gettext <other text>
    end)
  end
```
Seems verbose and I believe people will end up wrapping `Gettext.with_locale(MyApp.Gettext, locale, fun)` in their own function.

**Comment by whatyouhide** [comment]
Verbosity is without doubt the most negative thing about the `Gettext.*locale

**Comment by josevalim** [comment]
@lexmag I don't think folks will use with_locale like that, honestly. You will usually set the locale only once, at the beginning of your requests in your plug pipeline, and that's it. No need to swap the locale for every translation, with_locale is mostly for tests.

**Comment by josevalim** [comment]
:shipit:

**Comment by whatyouhide** [comment]
@josevalim yes, I agree setting the locale is something you do once and getting it is not a frequent operation probably. So you're in favour of `Gettext.*locale` functions versus `MyApp.Gettext.*locale` functions?

**Comment by lexmag** [comment]
@josevalim there is one real case I've encountered already – generating payload for push notifications. :smile:

**Comment by lexmag** [comment]
Anyway, I always for simplicity at starting, I'll just be wrapping it for now.

**Comment by lexmag** [comment]
I think I could avoid using `with_locale` in my case, the only thing that is needed then is `MyApp.Gettext.default_locale/0` – locale that can't be affected by any means.
Today it's doable with the following dance:

```elixir
defmodule MyApp.Gettext do
  use Gettext, otp_app: :my_app

  def default_locale() do
    __MODULE__.Defaults.__locale__
  end
end

defmodule MyApp.Gettext.Defaults do
  def __locale__() do
    unquote(Gettext.get_locale(MyApp.Gettext))
  end
end
```
@whatyouhide @josevalim what are your thoughts on having `default_locale/0`?

**Comment by whatyouhide** [comment]
@lexmag I'm probably missing something, but couldn't you do just:

```elixir
defmodule MyApp.Gettext do
  use Gettext, otp_app: :my_app

  def default_locale do
    unquote(Gettext.get_locale(__MODULE__))
  end
end
```

Btw, still assuming I'm not missing something, I would go with `Gettext.default_locale/1` given we can always retrieve the default locale with `Application.get_env/3`.

**Comment by lexmag** [comment]
@whatyouhide you can't do this without extra supporting module as `module MyApp.Gettext is not available` during `unquote`.

**Comment by whatyouhide** [comment]
@lexmag I see, but we could implement it as:

```elixir
defmodule Gettext do
  def default_locale(backend) do
    Application.get_env(backend.__gettext__(:otp_app), backend)) |> Keyword.get(:default_locale, "en")
  end
end
```

right?

**Comment by whatyouhide** [comment]
@lexmag or do you mean that people can mess with that using `Application.put_env/2`?

**Comment by lexmag** [comment]
If we go with `Gettext.default_locale/1` function I think `Application.get_env/2` is fine (extra `ets` call though).
If one mess with `Application.put_env/2` it's honestly not our problem.

**Comment by whatyouhide** [comment]
I 100% agree on that (people using `put_env/3`), that would have been my argument :). Personally I don't think the extra ets call is a problem as I don't think `default_locale/1` is a critical function. If it turns out to be, I'm more inclined to change all the `*locale` functions and move them to `MyApp.Gettext` rather than having `put_/get_/with_locale` in `Gettext` and `default_locale` in `MyApp.Gettext`. Wdyt?

**Comment by lexmag** [comment]
After some tryouts of v0.7 I believe we should think of backend modules as of main basic thing to interact with locales settings, i.e. having `*locale` functions in `MyApp.Gettext` feels truly natural.

**Comment by whatyouhide** [comment]
@lexmag I think I agree with that. Since they're per-module, it feels more natural to have them set by their module. Also, it's less verbose :smiley_cat: Let's hear what @josevalim thinks?

**Comment by josevalim** [comment]
:+1: for being per module although I would not add with_locale to the modules. Again, it is alright to be in Gettext as I most see it as a helper and not a first class API.

**Comment by whatyouhide** [comment]
Just one last doubt before moving the functions from `Gettext` to `MyApp.Gettext`: keeping them in `Gettext` has the benefit of easy documentation. Today, we don't generate docs for `*gettext` macros for example, so this is something we have to solve either way. Wdyt? /cc @josevalim @lexmag

**Comment by josevalim** [comment]
Sounds alright.

**Comment by lexmag** [comment]
Ping.

**Comment by whatyouhide** [comment]
oh god 0.0 pong, sorry!! Will work on this tomorrow, i promise :/

On Friday, 20 November 2015, Aleksei Magusev <notifications@github.com>
wrote:

> Ping.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/elixir-lang/gettext/pull/49#issuecomment-158519837>.
>


-- 

Andrea Leopardi
an.leopardi@gmail.com

**Comment by whatyouhide** [comment]
@lexmag hey there, so just a question: how would we go about documentation if we move all functions and macros from `Gettex` to `MyApp.Gettext`? I suppose we could generate them but I still like that people can look at the documentation for gettext without having to bring gettext + ex_doc in their projects :\

**Comment by josevalim** [comment]
@whatyouhide one common trick is to define them as `@callback` in the Gettext module. This way, you still have a space for documenting them and you can declare that each backend implements the `Gettext` module behaviour.

**Comment by josevalim** [comment]
@whatyouhide btw, what was the decision? will we add put_locale/get_locale to the backends as well?

**Comment by whatyouhide** [comment]
@josevalim I think we settled on *moving* `put_/get_locale` to the backends only, removing them from `Gettext`. Can you think of any downsides in doing that? We would leave `with_locale` in Gettext though, as as you said it's meant mostly for testing.

**Comment by lexmag** [comment]
@whatyouhide as José said, `@callback` definition would be the best option, thus we will have cross-references from implementation to behaviour, and decent docs for callbacks.

I think we agreed on:
* move `put_/get_locale` to the backends
* keep `with_locale` in `Gettext`
* introduce `default_locale/0` in the backends

**Comment by whatyouhide** [comment]
Alright, sounds like a plan. Working on it now :)

**Comment by josevalim** [comment]
I feel like we should keel `put_/get_locale` in `Gettext` only because it has the implementation of everything else. Also, I wouldn't introduce `default_locale/0` only because you can easily read it from the app config. Since all of this is generated, it is nice to keep those as small as possible.

**Comment by lexmag** [comment]
Okay, let's wait with these changes then, until more feedback comes. :smile:

**Pull request merged**